sockeye/lexical_constraints.py

```diff
@@ -0,0 +1,399 @@
+
+class ConstrainedHypothesis:
+    """
+    Represents a set of words and phrases that must appear in the output.
+    A constraint is of two types: sequence or non-sequence.
+    A non-sequence constraint is a single word and can therefore be followed by anything, whereas a sequence constraint must be followed by a particular word (the next word in the sequence).
+    This class also records which constraints have been met.
+
+    :param constraint_list: A list of zero or raw constraints (each represented as a list of integers).
+    :param eos_id: The end-of-sentence ID.
+    """
+    def __init__(self, constraint_list: RawConstraintList, eos_id: int) -> None:
+
+        # `constraints` records the words of the constraints, as a list (duplicates allowed).
+        # `is_sequence` is a parallel array that records, for each corresponding constraint,
+        self.constraints = []
+        self.is_sequence = []
+        for phrase in constraint_list:
+            self.constraints += phrase
+            self.is_sequence += [1] * len(phrase)
+            self.is_sequence[-1] = 0
+
+        self.eos_id = eos_id
+
+        # no constraints have been met
+        self.met = [False for x in self.constraints]
+        self.lastMet = -1
+
+    def __len__(self) -> int:
+        """
+        :return: The number of constraints.
+        """
+        return len(self.constraints)
+
+    def __str__(self) -> str:
+        s = []
+        for i, id in enumerate(self.constraints):
+            s.append(str(id) if self.met[i] is False else 'X')
+            if self.is_sequence[i]: s.append('->')
+        return ' '.join(s)
+
+    def size(self) -> int:
+        """
+        :return: the number of constraints
+        """
+        return len(self.constraints)
+
+    def numMet(self) -> int:
+        """
+        :return: the number of constraints that have been met.
+        """
+        return sum(self.met)
+
+    def numNeeded(self) -> int:
+        """
+        :return: the number of un-met constraints.
+        """
+        return self.size() - self.numMet()
+
+    def allowed(self) -> List[int]:
+        """
+        Returns the set of constrained words that could follow this one.
+        For unfinished phrasal constraints, it is the next word in the phrase.
+        In other cases, it is a list of unmet constraints.
+        If all constraints are met, an empty set is returned.
+
+        :return: The ID of the next required word, or -1 if any word can follow
+        """
+        items = []
+        # Add extensions of a started-but-incomplete sequential constraint
+        if self.lastMet != -1 and self.is_sequence[self.lastMet] == 1:
+            word_id = self.constraints[self.lastMet + 1]
+            if word_id != self.eos_id or self.numNeeded() == 1:
+                items.append(word_id)
+
+        # Add all constraints that aren't non-initial sequences
+        else:
+            for i, id in enumerate(self.constraints):
```

Comment:
`id` collides with the built-in function.  This could be `for i, _` since `id` doesn't appear to be used.

Reply:
done